imap/mboxlist.c

```diff
@@ -394,7 +403,13 @@ static void mboxlist_dbname_to_key(const char *dbname, size_t len,
 static void mboxlist_dbname_from_key(const char *key, size_t len,
                                      const char *userid, struct buf *dbname)
 {
-    if (userid && len >= 6 && !strncmp(key+1, "INBOX", 5)) {
+    assert(key[0] == KEY_TYPE_NAME);
+
+    if (userid
+        && len >= 6
+        && (!key[6] || key[6] == DB_HIERSEP_CHAR)
```

Comment:
Valgrind is unhappy about this. I think it's looking at key[6].

I'm guessing that cyrusdb_foreach is giving us a pointer to exactly 6 bytes, not 6 bytes plus 1 byte for null termination (which is why we have a key + keylen interface I imagine?)
```
==87065== Conditional jump or move depends on uninitialised value(s)
==87065==    at 0x4915E12: mboxlist_dbname_from_key (mboxlist.c:410)
==87065==    by 0x4917682: find_p (mboxlist.c:3584)
==87065==    by 0x4ACD55A: cyrusdb_forone (cyrusdb.c:295)
==87065==    by 0x491E600: mboxlist_do_find (mboxlist.c:4239)
==87065==    by 0x491FD58: mboxlist_findsubmulti_withp (mboxlist.c:5075)
==87065==    by 0x491FDD9: mboxlist_findsubmulti (mboxlist.c:5033)
==87065==    by 0x123C14: list_data (imapd.c:13865)
==87065==    by 0x123DBE: cmd_list (imapd.c:8545)
==87065==    by 0x13C769: cmdloop (imapd.c:2149)
==87065==    by 0x13F3E5: service_main (imapd.c:1264)
==87065==    by 0x153AA1: main (service.c:647)
```

Reply:
Oh yeah right, I see the issue.  I'm expecting `[6]` to either be a `'\0'` or a separator, and that's fine when `len > 6`, but my condition on the line above is `len >= 6` -- so there might not be a `[6]`.  Thanks!

Reply:
There's a couple of spots with this shape of thing, I'll have a close look at the other one too while I'm in there...

Reply:
And yes, you're right about the cyrusdb key/keylen, data/datalen interfaces using arbitrary memory buffers, not C strings.

Reply:
I've fixed both `mbname_dbname_to_key()` and `mbname_dbname_from_key()` to not read `[n]` when `len==n`, and the Shared and Subscriptions tests now pass cleanly for me with `--valgrind`